htlcswitch/switch.go

```diff
@@ -407,13 +416,15 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID,
 	// If the returned error is a duplicate add, then we can ignore it, as
 	// our HTLC was already forwarded by the switch.
 	if err := s.forward(packet); err != nil && err != ErrDuplicateAdd {
+		s.paymentIDMtx.Unlock(paymentID)
 		s.removePendingPayment(paymentID)
 		if err := s.control.Fail(htlc.PaymentHash); err != nil {
 			return zeroPreimage, err
 		}
 
 		return zeroPreimage, err
 	}
+	s.paymentIDMtx.Unlock(paymentID)
```

Comment:
Alternative is to `Unlock` directly after the call to `s.forward` and only then start checking the `err` value.

Reply:
Done!